lib/dependency_validator.dart

```diff
@@ -51,41 +52,47 @@ Future<Null> run() async {
       ?.toList();
   final excludes = configExcludes ?? <Glob>[];
   logger.fine('excludes:\n${bulletItems(excludes.map((g) => g.pattern))}\n');
-  final ignoredPackages = <String>[...commonBinaryPackages, ...config?.ignore ?? []];
+  final ignoredPackages = config?.ignore ?? <String>[];
   logger.fine('ignored packages:\n${bulletItems(ignoredPackages)}\n');
 
   // Read and parse the analysis_options.yaml in the current working directory.
   final optionsIncludePackage = getAnalysisOptionsIncludePackage();
 
   // Read and parse the pubspec.yaml in the current working directory.
-  final YamlMap pubspecYaml = loadYaml(File('pubspec.yaml').readAsStringSync());
+  final pubspec = Pubspec.parse(File('pubspec.yaml').readAsStringSync());
 
   // Extract the package name.
-  final packageName = pubspecYaml[nameKey];
+  final packageName = pubspec.name;
 
   logger.info('Validating dependencies for $packageName\n');
 
   // Find packages that provide executables.
   final packagesWithExecutables = Set<String>();
+  final packagesWithBuilders = Set<String>();
   final packageConfig = await findPackageConfig(Directory.current);
   for (final package in packageConfig.packages) {
+    // Search for builders
+    final buildConfig = await BuildConfig.fromBuildConfigDir(packageName, pubspec.dependencies.keys, package.root.path);
+    if (buildConfig.builderDefinitions.isNotEmpty) {
+      packagesWithBuilders.add(package.name);
+    }
```

Comment:
Since we have the fully parsed builder definitions, we should be able to go a little further than just determining whether the dependency defines any builders. It's possible that the builder(s) defined in a package may not be auto-applied–in which case the package might actually be unused–but it's also possible that the root package explicitly enables a builder in its own build.yaml.

Something like this should be able to determine whether the root package has a build.yaml config that references a builder from any given dependency:

```dart
  final rootBuildConfig = await BuildConfig.fromPackageDir('.');
  bool rootPackageReferencesDependencyInBuildYaml(String dependencyName) =>
    [
      ...rootBuildConfig.globalOptions.keys,
      for (final target in rootBuildConfig.buildTargets.values)
        ...target.builders.keys,
    ]
      .map(normalizeBuilderKeyUsage)
      .any((key) => key.startsWith('$dependencyName:');
```

And something like this should be able to determine whether a BuildConfig defines any builders that would be auto-applied to the root package:

```dart
  bool buildConfigDefinesAutoAppliedBuilder(BuildConfig buildConfig) =>
    buildConfig.builderDefinitions.values.any((def) => def.autoApply != AutoApply.none);
```

With those two, I think we could do something like this:

```suggestion
    if (buildConfigDefinesAutoAppliedBuilder(buildConfig) ||
        rootPackageReferencesDependencyInBuildYaml(package.name)) {
      packagesWithBuilders.add(package.name);
    }
```

And I guess maybe at that point it would be worth updating the list to be named `packagesWithConsumedBuilders` or something similar.